src/lib/utils/abstract_plugin.hpp

```diff
@@ -12,6 +12,21 @@ namespace opossum {
 #define EXPORT_PLUGIN(PluginName) \
   extern "C" AbstractPlugin* factory() { return new PluginName(); }
 
+// This macro allows the function `FunctionName` of the plugin called `PluginName` to be
+// externally callable. This is necessary to enable user-callable functions. As above,
+// it leads to plain C linkage. Use EXPORT in the implementation file of your plugin.
+#define EXPORT_USER_CALLABLE_FUNCTION(PluginName, FunctionName) \
+  /* Careful! We are getting a raw pointer from a smart (unique) pointer here. */ \
+  /* This could conflict with the intention of unique_ptr. It should not be    */ \
+  /* a problem, as long as we handle it responsible and let it go out of scope */ \
+  /* at the end of this function.                                              */ \
+  extern "C" void FunctionName ## _extern(const std::unique_ptr<AbstractPlugin>& plugin) { \
+    const auto plugin_raw = plugin.get(); \
```

Comment:
Unfortunately, I think there is no way around that.

Reply:
Do we have to use an `std::unique_ptr` here?

Reply:
The PluginManager stores plugins as unique_ptrs. I think in the beginning we thought that the PluginManager should have exclusive ownership and that there is no reason for shared_ptrs. We could change that but might go with the alternative implementation in the end anyway.

Reply:
With shared pointers, we can avoid a raw pointer object in this code location, a raw pointer cast, and the comments explaining why we use a raw pointer here.
The minimal overhead for reference counting with shared pointers should also be negligible. Am I missing some drawbacks of changing to shared pointers here?

Reply:
Technically, we could change to a `shared_ptr` but we chose the `unique_ptr` on purpose to declare exclusive ownership. The idea was that the PluginManager is not sharing plugin ownership because otherwise things could get nasty when unloading the shared library.